File: src/rvm/analysis/path_registry.rs
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

//! Path registry — maps access paths (e.g., `input.user.role`) to SMT variables.
//!
//! The path-based encoding is the central design choice: every access into `input`
//! (or symbolic `data`) discovered during translation gets a flat SMT variable named
//! by its access path. This makes model-to-JSON extraction trivial.

use alloc::format;
use alloc::string::String;
use alloc::string::ToString;
use alloc::vec;
use alloc::vec::Vec;

use std::collections::HashMap;

use regorus_smt::{SmtDecl, SmtExpr, SmtSort};

use super::types::ValueSort;

/// Map a `ValueSort` to the corresponding SMT sort.
#[allow(dead_code)]
fn value_sort_to_smt(sort: ValueSort) -> SmtSort {
    match sort {
        ValueSort::Bool => SmtSort::Bool,
        ValueSort::Int => SmtSort::Int,
        ValueSort::Real => SmtSort::Real,
        ValueSort::String => SmtSort::String,
        ValueSort::Unknown => SmtSort::String, // default
    }
}

/// A single path entry in the registry.
#[derive(Debug, Clone)]
pub struct PathEntry {
    /// The SMT boolean controlling whether this path is defined (present in input).
    pub defined: SmtExpr,
    /// The inferred sort for this path's value.
    pub sort: ValueSort,
    /// Whether all access components were literal (fully static path).
    pub is_static: bool,
    /// Instruction PCs that access this path (for coverage mapping).
    pub access_pcs: Vec<usize>,

    // The actual SMT variable (SmtExpr::Const(id)). One per sort, created on demand.
    pub(crate) bool_var: Option<SmtExpr>,
    pub(crate) int_var: Option<SmtExpr>,
    pub(crate) real_var: Option<SmtExpr>,
    pub(crate) str_var: Option<SmtExpr>,
}

/// Registry of all discovered access paths into `input` (and symbolic `data`).
#[allow(missing_debug_implementations)]
pub struct PathRegistry {
    paths: HashMap<String, PathEntry>,
    /// Counter for generating unique variable IDs.
    next_id: u32,
    /// All SMT constant declarations created by this registry.
    declarations: Vec<SmtDecl>,
}

impl PathRegistry {
    /// Create a new empty registry.
    pub fn new() -> Self {
        Self {
            paths: HashMap::new(),
            next_id: 0,
            declarations: Vec::new(),
        }
    }

    /// Allocate a fresh SMT constant, returning its `SmtExpr::Const(id)`.
    pub(crate) fn alloc_const(&mut self, name: &str, sort: SmtSort) -> SmtExpr {
        let id = self.next_id;
        self.next_id += 1;
        self.declarations.push(SmtDecl::Const {
            id,
            name: name.into(),
            sort,
        });
        SmtExpr::Const(id)
    }

    /// Get or create a path entry. If the path already exists, its sort may be
    /// refined (Unknown → concrete sort) but never changed between concrete sorts.
    pub fn get_or_create(
        &mut self,
        path: &str,
        sort: ValueSort,
        is_static: bool,
        pc: usize,
    ) -> &mut PathEntry {
        if self.paths.contains_key(path) {
            let entry = self.paths.get_mut(path).unwrap();
            entry.access_pcs.push(pc);
            if entry.sort == ValueSort::Unknown && sort != ValueSort::Unknown {
                entry.sort = sort;
            }
            return entry;
        }

        // Create new entry with a fresh `defined` boolean.
        let defined_name = format!("defined_{}", path);
        let defined = self.alloc_const(&defined_name, SmtSort::Bool);

        self.paths.insert(
            path.to_string(),
            PathEntry {
                defined,
                sort,
                is_static,
                access_pcs: vec![pc],
                bool_var: None,
                int_var: None,
                real_var: None,
                str_var: None,
            },
        );
        self.paths.get_mut(path).unwrap()
    }

    /// Check if any sort-specific variable already exists for this path.
    fn has_any_var(&self, path: &str) -> bool {
        if let Some(entry) = self.paths.get(path) {
            entry.bool_var.is_some()
                || entry.int_var.is_some()
                || entry.real_var.is_some()
                || entry.str_var.is_some()
        } else {
            false
        }
    }

    /// Get the SMT Bool variable for a path, creating it if needed.
    ///
    /// If a variable of a different sort already exists for this path,
    /// the new declaration uses a sort-disambiguated name (`path$Bool`)
    /// to avoid duplicate-name errors in SMT-LIB2 text rendering.
    pub fn get_bool(&mut self, path: &str) -> SmtExpr {
        self.get_or_create(path, ValueSort::Bool, true, 0);
        if let Some(var) = &self.paths.get(path).unwrap().bool_var {
            return var.clone();
        }
        let name = if self.has_any_var(path) {
            format!("{}$Bool", path)
        } else {
            path.to_string()
        };
        let var = self.alloc_const(&name, SmtSort::Bool);
        self.paths.get_mut(path).unwrap().bool_var = Some(var.clone());
        var
    }

    /// Get the SMT Int variable for a path, creating it if needed.
    pub fn get_int(&mut self, path: &str) -> SmtExpr {
        self.get_or_create(path, ValueSort::Int, true, 0);
        if let Some(var) = &self.paths.get(path).unwrap().int_var {
            return var.clone();
        }
        let name = if self.has_any_var(path) {
            format!("{}$Int", path)
        } else {
            path.to_string()
        };
        let var = self.alloc_const(&name, SmtSort::Int);
        self.paths.get_mut(path).unwrap().int_var = Some(var.clone());
        var
    }

    /// Get the SMT Real variable for a path, creating it if needed.
    pub fn get_real(&mut self, path: &str) -> SmtExpr {
        self.get_or_create(path, ValueSort::Real, true, 0);
        if let Some(var) = &self.paths.get(path).unwrap().real_var {
            return var.clone();
        }
        let name = if self.has_any_var(path) {
            format!("{}$Real", path)
        } else {
            path.to_string()
        };
        let var = self.alloc_const(&name, SmtSort::Real);
        self.paths.get_mut(path).unwrap().real_var = Some(var.clone());
        var
    }

    /// Get the SMT String variable for a path, creating it if needed.
    pub fn get_string(&mut self, path: &str) -> SmtExpr {
        self.get_or_create(path, ValueSort::String, true, 0);
        if let Some(var) = &self.paths.get(path).unwrap().str_var {
            return var.clone();
        }
        let name = if self.has_any_var(path) {
            format!("{}$String", path)
        } else {
            path.to_string()
        };
        let var = self.alloc_const(&name, SmtSort::String);
        self.paths.get_mut(path).unwrap().str_var = Some(var.clone());
        var
    }

    /// Get a path entry by name (immutable).
    pub fn get(&self, path: &str) -> Option<&PathEntry> {
        self.paths.get(path)
    }

    /// Iterate over all registered paths.
    pub fn iter(&self) -> impl Iterator<Item = (&str, &PathEntry)> {
        self.paths.iter().map(|(k, v)| (k.as_str(), v))
    }

    /// Number of registered paths.
    pub fn len(&self) -> usize {
        self.paths.len()
    }

    /// Is the registry empty?
    pub fn is_empty(&self) -> bool {
        self.paths.is_empty()
    }

    /// All SMT declarations created by this registry.
    pub fn declarations(&self) -> &[SmtDecl] {
        &self.declarations
    }

    /// Update the sort of a path and ensure the corresponding SMT variable exists.
    pub fn refine_sort(&mut self, path: &str, sort: ValueSort) {
        if let Some(entry) = self.paths.get_mut(path) {
            if entry.sort == ValueSort::Unknown {
                entry.sort = sort;
            }
        }
    }

    /// Pre-seed the registry with sort information derived from an example
    /// input value.  This walks the value recursively, mapping each leaf to
    /// its `ValueSort`.  Array elements are expanded up to `max_elements`
    /// indices so that symbolic witness paths like `input.servers[2].id`
    /// will already have a known sort when the translator encounters them.
    pub fn seed_sorts_from_value(
        &mut self,
        prefix: &str,
        value: &super::super::super::value::Value,
        max_elements: usize,
    ) {
        use super::super::super::value::Value;
        match value {
            Value::Object(obj) => {
                for (k, v) in obj.iter() {
                    let key_str = match k {
                        Value::String(s) => s.to_string(),
                        other => format!("{:?}", other),
                    };
                    let child_path = format!("{}.{}", prefix, key_str);
                    self.seed_sorts_from_value(&child_path, v, max_elements);
                }
            }
            Value::Array(arr) => {
                // Seed sorts for indices 0..max_elements using the types
                // observed in the example array elements.
                for idx in 0..max_elements {
                    if arr.is_empty() {
                        break;
                    }
                    let elem = &arr[idx % arr.len()];
                    let child_path = format!("{}[{}]", prefix, idx);
                    self.seed_sorts_from_value(&child_path, elem, max_elements);
                }
            }
            Value::Bool(_) => {
                // Seed the sort AND create the SMT variable so it's available
                // for both constraint generation and model extraction.
                self.get_bool(prefix);
            }
            Value::Number(n) => {
                if n.as_f64().is_some() && n.as_i64().is_none() {
                    self.get_real(prefix);
                } else {
                    self.get_int(prefix);
                }
            }
            Value::String(_) => {
                self.get_string(prefix);
            }
            _ => {
                // Null, Undefined, Set — skip.
            }
        }
    }

    /// Get the sort registered for a path, if it exists.
    pub fn get_sort(&self, path: &str) -> Option<ValueSort> {
        self.paths.get(path).map(|e| e.sort)
    }

    /// Get an SMT variable for a path with the given sort, creating if needed.
    /// Returns `None` if the path doesn't exist.
    pub fn get_var_for_sort(
        &mut self,
        path: &str,
        sort: ValueSort,
    ) -> Option<super::types::SymValue> {
        use super::types::SymValue;

        // Ensure the entry exists
        if !self.paths.contains_key(path) {
            return None;
        }

        match sort {
            ValueSort::Bool => Some(SymValue::Bool(self.get_bool(path))),
            ValueSort::Int => Some(SymValue::Int(self.get_int(path))),
            ValueSort::Real => Some(SymValue::Real(self.get_real(path))),
            ValueSort::String => Some(SymValue::Str(self.get_string(path))),
            ValueSort::Unknown => {
                // Default to String for unknown sorts (most general for Rego).
                Some(SymValue::Str(self.get_string(path)))
            }
        }
    }
}

File: tools/z3analyze/path_registry.py
# Copyright (c) Microsoft Corporation.
# Licensed under the MIT License.
"""Path registry: flat map of input paths → Z3 variables with lazy creation."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Optional, Iterator

import z3

from .types import ValueSort, SymValue, sort_from_value


# ---------------------------------------------------------------------------
# PathEntry
# ---------------------------------------------------------------------------

@dataclass
class PathEntry:
    """One entry in the path registry — a single input path like 'input.user.role'."""

    defined: z3.BoolRef  # Z3 bool: is this path present?
    sort: ValueSort = ValueSort.Unknown
    is_static: bool = True
    access_pcs: list[int] = field(default_factory=list)

    # Lazily-created Z3 variables (one per sort)
    _bool_var: Optional[z3.BoolRef] = field(default=None, repr=False)
    _int_var: Optional[z3.ArithRef] = field(default=None, repr=False)
    _real_var: Optional[z3.ArithRef] = field(default=None, repr=False)
    _str_var: Optional[z3.SeqRef] = field(default=None, repr=False)


# ---------------------------------------------------------------------------
# PathRegistry
# ---------------------------------------------------------------------------

class PathRegistry:
    """
    Flat map of path strings to PathEntry.

    Paths like ``input.user.role`` or ``input.servers[0].id`` each get:
    - a ``defined`` Z3 Bool (created eagerly)
    - typed Z3 variables (created lazily on first access)
    """

    def __init__(self):
        self._paths: dict[str, PathEntry] = {}
        self._next_id: int = 0

    # ----- core access -----

    def get_or_create(
        self,
        path: str,
        sort: ValueSort = ValueSort.Unknown,
        is_static: bool = True,
        pc: int = 0,
    ) -> PathEntry:
        """Get or create a PathEntry for *path*, refining sort if needed."""
        entry = self._paths.get(path)
        if entry is not None:
            entry.access_pcs.append(pc)
            # refine sort: Unknown → concrete, but never overwrite concrete
            if entry.sort == ValueSort.Unknown and sort != ValueSort.Unknown:
                entry.sort = sort
            return entry

        # new entry
        self._next_id += 1
        defined_var = z3.Bool(f"defined_{path}")
        entry = PathEntry(defined=defined_var, sort=sort, is_static=is_static, access_pcs=[pc])
        self._paths[path] = entry
        return entry

    def get(self, path: str) -> Optional[PathEntry]:
        return self._paths.get(path)

    def __contains__(self, path: str) -> bool:
        return path in self._paths

    def __len__(self) -> int:
        return len(self._paths)

    def items(self) -> Iterator[tuple[str, PathEntry]]:
        return iter(self._paths.items())

    # ----- typed Z3 variable accessors (lazy creation) -----

    def get_bool(self, path: str) -> z3.BoolRef:
        entry = self.get_or_create(path, ValueSort.Bool)
        if entry._bool_var is None:
            entry._bool_var = z3.Bool(path)
            entry.sort = ValueSort.Bool
        return entry._bool_var

    def get_int(self, path: str) -> z3.ArithRef:
        entry = self.get_or_create(path, ValueSort.Int)
        if entry._int_var is None:
            entry._int_var = z3.Int(path)
            entry.sort = ValueSort.Int
        return entry._int_var

    def get_real(self, path: str) -> z3.ArithRef:
        entry = self.get_or_create(path, ValueSort.Real)
        if entry._real_var is None:
            entry._real_var = z3.Real(path)
            entry.sort = ValueSort.Real
        return entry._real_var

    def get_string(self, path: str) -> z3.SeqRef:
        entry = self.get_or_create(path, ValueSort.String)
        if entry._str_var is None:
            entry._str_var = z3.String(path)
            entry.sort = ValueSort.String
        return entry._str_var

    def get_var_for_sort(self, path: str, sort: ValueSort) -> Optional[SymValue]:
        """Return a SymValue wrapping the Z3 variable for the given sort, or None."""
        if sort == ValueSort.Bool:
            return SymValue.from_z3_bool(self.get_bool(path))
        if sort == ValueSort.Int:
            return SymValue.from_z3_int(self.get_int(path))
        if sort == ValueSort.Real:
            return SymValue.from_z3_real(self.get_real(path))
        if sort == ValueSort.String:
            return SymValue.from_z3_string(self.get_string(path))
        return None

    def refine_sort(self, path: str, sort: ValueSort):
        entry = self._paths.get(path)
        if entry and entry.sort == ValueSort.Unknown:
            entry.sort = sort

    def get_sort(self, path: str) -> Optional[ValueSort]:
        entry = self._paths.get(path)
        return entry.sort if entry else None

    # ----- seeding from a concrete JSON value -----

    def seed_sorts_from_value(self, prefix: str, value, max_elements: int = 4):
        """
        Recursively walk a concrete value (parsed JSON) to pre-populate
        the registry with correct sorts before symbolic translation.
        """
        if isinstance(value, dict):
            for k, v in value.items():
                self.seed_sorts_from_value(f"{prefix}.{k}", v, max_elements)
        elif isinstance(value, list):
            if not value:
                return
            for idx in range(max_elements):
                elem = value[idx % len(value)]
                self.seed_sorts_from_value(f"{prefix}[{idx}]", elem, max_elements)
        elif isinstance(value, bool):
            self.get_bool(prefix)
        elif isinstance(value, int):
            self.get_int(prefix)
        elif isinstance(value, float):
            self.get_real(prefix)
        elif isinstance(value, str):
            self.get_string(prefix)
        # None and other types: skip
